include/nbl/builtin/hlsl/workgroup/shared_scan.hlsl

```diff
@@ -0,0 +1,166 @@
+        // the first gl_SubgroupSz^2 invocations will be processed by the first subgroup and so on.
+        // Consequently, those first gl_SubgroupSz^2 invocations will store their results on gl_SubgroupSz scratch slots 
+        // and the next level will follow the same + the previous as an `offset`.
+        
+        scanLoadIndex = gl_LocalInvocationIndex;
+        const uint loadStoreIndexDiff = scanLoadIndex - glsl::gl_SubgroupID();
+        
+        bool participate = gl_LocalInvocationIndex <= lastInvocationInLevel;
+        while(lastInvocationInLevel >= glsl::gl_SubgroupSize() * glsl::gl_SubgroupSize())
+        {
+            if(participate)
+            {
+                if (any(bool2(gl_LocalInvocationIndex == lastInvocationInLevel, isLastSubgroupInvocation)))
+                { // only invocations that have the final value of the subgroupOp (inclusive scan) store their results
+                    sharedAccessor.main.set(scanLoadIndex - loadStoreIndexDiff, scan); // For gl_SubgroupSz = 64, first 4095 invocations store index is [0,63], 4096-8191 [64,127] etc.
+                }
+            }
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
+            if(participate)
+            {
+                const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
+                scan = subgroupOp(prevLevelScan);
+                sharedAccessor.main.set(scanLoadIndex, scan);
+                scanLoadIndex += lastInvocationInLevel + 1u;
+            }
+        }
+        if(lastInvocationInLevel >= glsl::gl_SubgroupSize())
+        {
+            if(participate)
+            {
+                if(any(bool2(gl_LocalInvocationIndex == lastInvocationInLevel, isLastSubgroupInvocation))) {
+                    sharedAccessor.main.set(scanLoadIndex - loadStoreIndexDiff, scan);
+                }
+            }
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
+            if(participate)
+            {
+                const uint prevLevelScan = sharedAccessor.main.get(scanLoadIndex);
+                scan = subgroupOp(prevLevelScan);
+            }
+        }
+        lastLevelScan = scan; // only invocations of gl_LocalInvocationIndex < gl_SubgroupSize will have correct values, rest will have garbage
+    }
+};
+
+template<typename T, class Binop, class SubgroupScanOp, class SharedAccessor, bool isExclusive>
+struct Scan
+{
+    T identity;
+    Reduce<T, SubgroupScanOp, SharedAccessor> reduce;
+
+    static Scan create(uint itemCount, T identity)
+    {
+        Scan<T, Binop, SubgroupScanOp, SharedAccessor, isExclusive> scan;
+        scan.identity = identity;
```

Comment:
you don't need the bionop unless `isExclusive` also juse use it as `Binop::identity` where its needed no need to fling it around all the code